File: node-src/domain/zcc-collector-catalog.ts
import embeddedCatalogJson from "../../catalogs/zcc-collector-catalog.v1.json" with { type: "json" };

import { isJsonRecord, pythonJsonEqual } from "../json/python-equality.js";
import { sameStringSequence, sortedStrings } from "../json/python-compatible.js";
import { snapshotPlainJsonGraph } from "../json/supported-json-graph.js";
import { ProcessFailure } from "./errors.js";

const RESOURCE_TYPES = [
  "zcc_device_cleanup",
  "zcc_failopen_policy",
  "zcc_forwarding_profile",
  "zcc_trusted_network",
  "zcc_web_privacy",
] as const;

const SOURCE_FILES = [
  "engine/collectors/rest/__init__.py",
  "packs/_shared/zscaler/collector.py",
  "packs/zcc/collector.py",
  "packs/zcc/pack.json",
  "packs/zcc/registry.json",
] as const;

const SOURCES_SHA256 =
  "d4b8cbef8294e8cb7fd5b17b6efb120b5f8bdc09de8c10e506763547748b11fc";

const RESOURCE_CONTRACTS = {
  zcc_device_cleanup: {
    envelope: null,
    page_size: null,
    pagination: "single",
    path: "zcc/papi/public/v1/getDeviceCleanupInfo",
  },
  zcc_failopen_policy: {
    envelope: null,
    page_size: 1000,
    pagination: "zia",
    path: "zcc/papi/public/v1/webFailOpenPolicy/listByCompany",
  },
  zcc_forwarding_profile: {
    envelope: null,
    page_size: 1000,
    pagination: "zia",
    path: "zcc/papi/public/v1/webForwardingProfile/listByCompany",
  },
  zcc_trusted_network: {
    envelope: "trustedNetworkContracts",
    page_size: 1000,
    pagination: "zia",
    path: "zcc/papi/public/v1/webTrustedNetwork/listByCompany",
  },
  zcc_web_privacy: {
    envelope: null,
    page_size: null,
    pagination: "single",
    path: "zcc/papi/public/v1/getWebPrivacyInfo",
  },
} as const;

export type ZccCollectorResourceType = typeof RESOURCE_TYPES[number];

export interface ZccCollectorCatalogResource {
  readonly envelope: "trustedNetworkContracts" | null;
  readonly method: "GET";
  readonly page_size: 1000 | null;
  readonly pagination: "single" | "zia";
  readonly path: string;
  readonly type: ZccCollectorResourceType;
}

export interface ZccCollectorCatalog {
  readonly kind: "infrawright.zcc_collector_catalog";
  readonly oneapi: {
    readonly audience: "https://api.zscaler.com";
    readonly cloud_gateway_template: "https://api.{cloud}.zsapi.net";
    readonly cloud_token_host_template: "https://{vanity}.zslogin{cloud}.net";
    readonly mode: "oneapi";
    readonly production_gateway: "https://api.zsapi.net";
    readonly production_token_host_template: "https://{vanity}.zslogin.net";
    readonly token_path: "/oauth2/v1/token";
  };
  readonly product: "zcc";
  readonly provider: {
    readonly source: "zscaler/zcc";
    readonly version: "0.1.0-beta.1";
  };
  readonly resources: readonly ZccCollectorCatalogResource[];
  readonly schema_version: 1;
  readonly source_files: readonly string[];
  readonly sources_sha256: string;
}

function invalidCatalog(message: string): never {
  throw new ProcessFailure({
    category: "domain",
    code: "INVALID_ZCC_COLLECTOR_CATALOG",
    message,
  });
}

function exactKeys(value: object, expected: readonly string[]): boolean {
  return sameStringSequence(sortedStrings(Object.keys(value)), expected);
}

function exactStrings(value: unknown, expected: readonly string[]): boolean {
  return Array.isArray(value)
    && value.every((item) => typeof item === "string")
    && sameStringSequence(value, expected);
}

function validResource(
  value: unknown,
  expectedType: ZccCollectorResourceType,
): value is ZccCollectorCatalogResource {
  if (!isJsonRecord(value) || !exactKeys(value, [
    "envelope", "method", "page_size", "pagination", "path", "type",
  ])) {
    return false;
  }
  const expected = RESOURCE_CONTRACTS[expectedType];
  return value.type === expectedType
    && value.method === "GET"
    && value.path === expected.path
    && value.envelope === expected.envelope
    && value.pagination === expected.pagination
    && value.page_size === expected.page_size;
}

function validatedCatalog(candidate: unknown): ZccCollectorCatalog {
  const snapshot = snapshotPlainJsonGraph(candidate, { maxDepth: 32 });
  if (!snapshot.ok || !isJsonRecord(snapshot.value) || !exactKeys(snapshot.value, [
    "kind", "oneapi", "product", "provider", "resources", "schema_version",
    "source_files", "sources_sha256",
  ])) {
    return invalidCatalog("ZCC collector catalog has an invalid top-level shape");
  }
  const value = snapshot.value;
  const oneapi = value.oneapi;
  const provider = value.provider;
  if (
    value.kind !== "infrawright.zcc_collector_catalog"
    || value.schema_version !== 1
    || value.product !== "zcc"
    || !isJsonRecord(oneapi)
    || !exactKeys(oneapi, [
      "audience", "cloud_gateway_template", "cloud_token_host_template", "mode",
      "production_gateway", "production_token_host_template", "token_path",
    ])
    || oneapi.audience !== "https://api.zscaler.com"
    || oneapi.cloud_gateway_template !== "https://api.{cloud}.zsapi.net"
    || oneapi.cloud_token_host_template !== "https://{vanity}.zslogin{cloud}.net"
    || oneapi.mode !== "oneapi"
    || oneapi.production_gateway !== "https://api.zsapi.net"
    || oneapi.production_token_host_template !== "https://{vanity}.zslogin.net"
    || oneapi.token_path !== "/oauth2/v1/token"
    || !isJsonRecord(provider)
    || !exactKeys(provider, ["source", "version"])
    || provider.source !== "zscaler/zcc"
    || provider.version !== "0.1.0-beta.1"
    || !Array.isArray(value.resources)
    || value.resources.length !== RESOURCE_TYPES.length
    || !value.resources.every((resource, index) => {
      const expected = RESOURCE_TYPES[index];
      return expected !== undefined && validResource(resource, expected);
    })
    || !exactStrings(value.source_files, SOURCE_FILES)
    || value.sources_sha256 !== SOURCES_SHA256
  ) {
    return invalidCatalog("ZCC collector catalog is malformed");
  }
  return snapshot.value as unknown as ZccCollectorCatalog;
}

const EMBEDDED_CATALOG = validatedCatalog(embeddedCatalogJson);

/** Return the private, immutable, exact-five ZCC collector contract. */
export function loadZccCollectorCatalog(): ZccCollectorCatalog {
  return EMBEDDED_CATALOG;
}

/** Accept only a semantic copy of the catalog embedded in this build. */
export function requireSupportedZccCollectorCatalog(
  candidate: unknown,
): ZccCollectorCatalog {
  const catalog = validatedCatalog(candidate);
  if (!pythonJsonEqual(catalog, EMBEDDED_CATALOG)) {
    throw new ProcessFailure({
      category: "domain",
      code: "UNSUPPORTED_ZCC_COLLECTOR_CATALOG",
      message: "collector requires the supported embedded ZCC catalog",
    });
  }
  return EMBEDDED_CATALOG;
}

export function zccCollectorResource(
  resourceType: string,
): ZccCollectorCatalogResource {
  const resource = EMBEDDED_CATALOG.resources.find((entry) => {
    return entry.type === resourceType;
  });
  if (resource === undefined) {
    throw new ProcessFailure({
      category: "domain",
      code: "UNSUPPORTED_ZCC_COLLECTOR_RESOURCE",
      message: "collector supports only the reviewed ZCC resource cohort",
    });
  }
  return resource;
}

File: tools/zcc_collector_catalog.py
"""Build the private, closed ZCC Node collector catalog.

This authoring-time tool binds the Node collector kernel to the exact five
fetch-backed ZCC resources and the committed Python OneAPI collector seams.
Node never loads pack files at runtime.
"""
import argparse
import hashlib
import json
import os
import sys


ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
CATALOG_KIND = "infrawright.zcc_collector_catalog"
SCHEMA_VERSION = 1
PRODUCT = "zcc"
PROVIDER_SOURCE = "zscaler/zcc"
PROVIDER_VERSION = "0.1.0-beta.1"
AUDIENCE = "https://api.zscaler.com"
PAGE_SIZE = 1000
COHORT_RESOURCES = (
    "zcc_device_cleanup",
    "zcc_failopen_policy",
    "zcc_forwarding_profile",
    "zcc_trusted_network",
    "zcc_web_privacy",
)
SOURCE_FILES = (
    "engine/collectors/rest/__init__.py",
    "packs/_shared/zscaler/collector.py",
    "packs/zcc/collector.py",
    "packs/zcc/pack.json",
    "packs/zcc/registry.json",
)
RESOURCE_KEYS = frozenset(["fetch", "generate", "product"])
FETCH_KEYS = frozenset(["envelope", "pagination", "path"])
RESOURCE_CONTRACTS = {
    "zcc_device_cleanup": {
        "envelope": None,
        "page_size": None,
        "pagination": "single",
        "path": "zcc/papi/public/v1/getDeviceCleanupInfo",
    },
    "zcc_failopen_policy": {
        "envelope": None,
        "page_size": PAGE_SIZE,
        "pagination": "zia",
        "path": "zcc/papi/public/v1/webFailOpenPolicy/listByCompany",
    },
    "zcc_forwarding_profile": {
        "envelope": None,
        "page_size": PAGE_SIZE,
        "pagination": "zia",
        "path": "zcc/papi/public/v1/webForwardingProfile/listByCompany",
    },
    "zcc_trusted_network": {
        "envelope": "trustedNetworkContracts",
        "page_size": PAGE_SIZE,
        "pagination": "zia",
        "path": "zcc/papi/public/v1/webTrustedNetwork/listByCompany",
    },
    "zcc_web_privacy": {
        "envelope": None,
        "page_size": None,
        "pagination": "single",
        "path": "zcc/papi/public/v1/getWebPrivacyInfo",
    },
}


def _read_json(relative_path):
    with open(os.path.join(ROOT, relative_path), encoding="utf-8") as f:
        return json.load(f)


def _source_digest():
    digest = hashlib.sha256()
    for relative_path in SOURCE_FILES:
        path = os.path.join(ROOT, relative_path)
        if not os.path.isfile(path):
            raise ValueError("collector source is missing: %s" % relative_path)
        with open(path, "rb") as f:
            content = f.read()
        digest.update(relative_path.encode("utf-8"))
        digest.update(b"\0")
        digest.update(content)
        digest.update(b"\0")
    return digest.hexdigest()


def _resource(resource_type, entry):
    if not isinstance(entry, dict):
        raise ValueError("registry is missing %s" % resource_type)
    unknown_resource_keys = sorted(set(entry) - RESOURCE_KEYS)
    if unknown_resource_keys:
        raise ValueError(
            "%s registry metadata has unsupported key %s"
            % (resource_type, unknown_resource_keys[0])
        )
    if entry.get("product") != PRODUCT or entry.get("generate") is not True:
        raise ValueError("%s is not an enabled ZCC resource" % resource_type)
    fetch = entry.get("fetch")
    if not isinstance(fetch, dict):
        raise ValueError("%s is not fetch-backed" % resource_type)
    unknown_fetch_keys = sorted(set(fetch) - FETCH_KEYS)
    if unknown_fetch_keys:
        raise ValueError(
            "%s fetch metadata has unsupported key %s"
            % (resource_type, unknown_fetch_keys[0])
        )
    pagination = fetch.get("pagination")
    if pagination not in ("single", "zia"):
        raise ValueError(
            "%s has unsupported pagination %r"
            % (resource_type, pagination)
        )
    path = fetch.get("path")
    if not isinstance(path, str) or not path or path.startswith("/"):
        raise ValueError("%s has an invalid OneAPI path" % resource_type)
    envelope = fetch.get("envelope")
    if resource_type == "zcc_trusted_network":
        if envelope != "trustedNetworkContracts":
            raise ValueError("trusted-network envelope drifted")
    elif envelope is not None:
        raise ValueError("%s unexpectedly declares an envelope" % resource_type)
    resource = {
        "envelope": envelope,
        "method": "GET",
        "page_size": PAGE_SIZE if pagination == "zia" else None,
        "pagination": pagination,
        "path": path,
        "type": resource_type,
    }
    expected = dict(RESOURCE_CONTRACTS[resource_type])
    expected.update({"method": "GET", "type": resource_type})
    if resource != expected:
        raise ValueError("%s fetch contract drifted" % resource_type)
    return resource


def build_catalog():
    manifest = _read_json("packs/zcc/pack.json")
    registry = _read_json("packs/zcc/registry.json")
    if not isinstance(registry, dict):
        raise ValueError("ZCC registry must be an object")
    fetch_backed = sorted(
        resource_type for resource_type, entry in registry.items()
        if isinstance(entry, dict) and "fetch" in entry
    )
    if fetch_backed != list(COHORT_RESOURCES):
        raise ValueError(
            "ZCC fetch cohort must remain exactly %s"
            % ", ".join(COHORT_RESOURCES)
        )
    provider = (manifest.get("provider_sources") or {}).get(PRODUCT)
    version = manifest.get("pin")
    if (
            provider != PROVIDER_SOURCE
            or version != PROVIDER_VERSION
            or manifest.get("provider_prefixes") != {"zcc_": "zcc"}
            or manifest.get("requires_shared") != ["zscaler"]):
        raise ValueError("ZCC pack provider metadata drifted")
    return {
        "kind": CATALOG_KIND,
        "oneapi": {
            "audience": AUDIENCE,
            "cloud_gateway_template": "https://api.{cloud}.zsapi.net",
            "cloud_token_host_template": (
                "https://{vanity}.zslogin{cloud}.net"
            ),
            "mode": "oneapi",
            "production_gateway": "https://api.zsapi.net",
            "production_token_host_template": (
                "https://{vanity}.zslogin.net"
            ),
            "token_path": "/oauth2/v1/token",
        },
        "product": PRODUCT,
        "provider": {
            "source": provider,
            "version": version,
        },
        "resources": [
            _resource(resource_type, registry.get(resource_type))
            for resource_type in COHORT_RESOURCES
        ],
        "schema_version": SCHEMA_VERSION,
        "source_files": list(SOURCE_FILES),
        "sources_sha256": _source_digest(),
    }


def render_catalog():
    return json.dumps(build_catalog(), indent=2, sort_keys=True) + "\n"


def main(argv=None):
    parser = argparse.ArgumentParser(
        prog="python3 tools/zcc_collector_catalog.py"
    )
    output = parser.add_mutually_exclusive_group(required=True)
    output.add_argument("--out")
    output.add_argument("--check")
    args = parser.parse_args(argv)
    try:
        text = render_catalog()
        if args.check:
            with open(args.check, encoding="utf-8") as f:
                actual = f.read()
            if actual != text:
                sys.stderr.write(
                    "error: ZCC collector catalog is stale: %s\n"
                    % args.check
                )
                return 1
            return 0
        with open(args.out, "w", encoding="utf-8") as f:
            f.write(text)
    except (IOError, OSError, KeyError, TypeError, ValueError) as exc:
        sys.stderr.write("error: %s\n" % exc)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
